ARGO/include/BehaviourTree.h

```diff
@@ -0,0 +1,172 @@
+#pragma once
+#include <Entity.h>
+
+////----------------------------This is the data we will need for our behaviour to function.----------------------------
+const int BOT_CAN_SEE_ENEMY_DISTANCE = Utilities::TILE_SIZE * 8;
+const int BOT_DISTANCE_TO_CARE_ABOUT_GOAL_SCALER = Utilities::TILE_SIZE * 2;
+const int BOT_DEFAUL_MOVE_TOWARD_GOAL_WEIGHT = 5;
+
+struct EnemyData
+{
+	Entity* enemy;
+	float distance;
+	int nearbyEnemies;
+};
+struct ClosestLeaderData
+{
+	Entity* entity;
+	float distance;
+};
+struct ClosestPickupData
+{
+	Entity* entity;
+	float distance;
+};
+struct GoalData
+{
+	Entity* entity;
+	float distance;
+};
+
+//----------------------------These are the basic classes needed for the behaviour tree to function----------------------------
+class WeightedNode 
+{
+public:
+	virtual bool run(Entity& t_entity) = 0;
+	virtual float getWeight() = 0;
+};
+
+class WeightedCompositeNode : public WeightedNode 
+{
+private:
+	std::vector<WeightedNode*> children;
+public:
+	const std::vector<WeightedNode*>& getChildren() const { return children; }
+	void addChild(WeightedNode* child) { children.emplace_back(child); }
+};
+
+class WeightedSelector : public WeightedCompositeNode 
+{
+public:
+	virtual bool run(Entity& t_entity) override
+	{
+		WeightedNode* bestNode = nullptr;
+		float bestWeight = -1.0f;
+		for (WeightedNode* child : getChildren()) 
+		{
+			float weight = child->getWeight();
+			if (weight > bestWeight)
+			{
+				bestNode = child;
+				bestWeight = weight;
+			}
+		}
+		if (bestNode)
+		{
+			bestNode->run(t_entity);
+			return true;
+		}
+		return false;
+	}
+	virtual float getWeight() override 
+	{
+		return 0.0f;
+	}
+};
+
+//----------------------------These are the the nodes that represent the different behaviour the tree is capable of----------------------------
+class RetreatBehaviour : public WeightedNode {
+private:
+	EnemyData* m_data;
+public:
+	RetreatBehaviour(EnemyData* t_data) : m_data(t_data) {}
+	virtual bool run(Entity& t_entity) override
+	{
+		glm::vec2 pos = static_cast<TransformComponent*>(t_entity.getComponent(ComponentType::Transform))->getPos();
+		ForceComponent* forceComp = static_cast<ForceComponent*>(t_entity.getComponent(ComponentType::Force));
+		glm::vec2 enemyPos = static_cast<TransformComponent*>(m_data->enemy->getComponent(ComponentType::Transform))->getPos();
+
+		if (forceComp)
+		{
+			forceComp->addForce(glm::normalize(pos - enemyPos));
+			return true;
+		}
+		return false;
+	}
+	virtual float getWeight() override
+	{
+		return (1 - m_data->distance / (BOT_CAN_SEE_ENEMY_DISTANCE * BOT_CAN_SEE_ENEMY_DISTANCE)) * 10;
```

Comment:
magic numbers 1 and 10

Reply:
At least have a comment explaining wtf are you doing here @NerakaiStudent 👍